storage/s3/src/main/java/io/aiven/kafka/tieredstorage/storage/s3/S3UploadOutputStream.java

```diff
@@ -138,13 +154,24 @@ public void close() throws IOException {
                 try {
                     partBuffer.position(0);
                     partBuffer.limit(lastPosition);
-                    flushBuffer(partBuffer.slice(), lastPosition);
+                    flushBuffer(partBuffer.slice(), lastPosition, multiPartUploadStarted());
                 } catch (final RuntimeException e) {
-                    log.error("Failed to upload last part {}, aborting transaction", uploadId, e);
-                    abortUpload();
+                    if (multiPartUploadStarted()) {
+                        log.error("Failed to upload last part {}, aborting transaction", uploadId, e);
+                        abortUpload();
+                    } else {
+                        closed = true;
```

Comment:
I realized that I asked you to remove the `closed = true` as part of the upload single file, while complete/abort MPU operation have this side-effect inside.
Now that there are two ways to complete the operation, I think it's worth moving those side-effects out and handle them here. So in this case you can clse outside the conditional (maybe before the try-catch? so we don't need to change the flag before throw and after try-catch?)

Reply:
@jeqo  done. you can help to check the latest commit. I set the close status to avoid the exception throw. Maybe it should be better and simple the code logic.